platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java

```diff
@@ -203,14 +204,39 @@ public Set<Object> getFrom() {
     public void setFromAnyValue(Object object) {
         // Currently we support just FileCollection for Groovy assign, so first try to cast to FileCollection
         FileCollectionInternal fileCollection = Cast.castNullable(FileCollectionInternal.class, Cast.castNullable(FileCollection.class, object));
+        checkNoSelfReferences(fileCollection);
+
+        setFrom(Cast.castNullable(FileCollection.class, object));
+    }
+
+    /**
+     * Verifies there is no self-referencing on value assignment.
+     * <p>
+     * Cover cases like: a += b, a = (a + b), a -= b, a = (a - b)
+     */
+    private void checkNoSelfReferences(FileCollectionInternal value) {
+        value.visitStructure(new FileCollectionStructureVisitor() {
+            private boolean isSubtracting = false;
 
-        // Don't allow a += b or a = (a + b), this is not support
-        fileCollection.visitStructure(new FileCollectionStructureVisitor() {
             @Override
-            public boolean startVisit(FileCollectionInternal.Source source, FileCollectionInternal fileCollection) {
+            public boolean startVisit(Source source, FileCollectionInternal fileCollection) {
                 if (DefaultConfigurableFileCollection.this == fileCollection) {
-                    throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Use the from() method to add to a ConfigurableFileCollection.");
+                    if (isSubtracting) {
+                        throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Avoid usage of the minus() or minus-assign operators.");
+                    } else {
+                        throw new UnsupportedOperationException("Self-referencing ConfigurableFileCollections are not supported. Use the from() method to add to a ConfigurableFileCollection.");
+                    }
+                }
+
+                if (fileCollection instanceof SubtractingFileCollection) {
+                    isSubtracting = true;
+                    SubtractingFileCollection subtraction = (SubtractingFileCollection) fileCollection;
+                    subtraction.getLeft().visitStructure(this);
+                    if (subtraction.getRight() instanceof FileCollectionInternal) {
+                        ((FileCollectionInternal) subtraction.getRight()).visitStructure(this);
+                    }
```

Comment:
💅 `return false`? We don't want to descent into children again even if `SubtractingFileCollection` becomes `CompositeFileCollection` later.